subsys/bluetooth/mesh/shell/shell_onoff_cli.c

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <stdlib.h>
+#include <bluetooth/mesh/models.h>
+#include <shell/shell.h>
+
+#include "mesh/net.h"
+#include "mesh/access.h"
+#include "shell_utils.h"
+
+static struct bt_mesh_model *mod;
+
+static int shell_onoff_get(const struct shell *shell, size_t argc, char *argv[])
+{
+	if (!mod) {
+		if (!shell_model_first_get(BT_MESH_MODEL_ID_GEN_ONOFF_CLI, &mod)) {
+			return -ENODEV;
+		}
+	}
+
+	struct bt_mesh_onoff_cli *cli = mod->user_data;
+	struct bt_mesh_onoff_status rsp;
+
+	int err = bt_mesh_onoff_cli_get(cli, NULL, &rsp);
+
+	if (!err) {
+		shell_print(shell, "Present val: %d, target val: %d, rem time: %d",
+			    rsp.present_on_off, rsp.target_on_off, rsp.remaining_time);
+	}
+
+	return err;
+}
+
+static int onoff_set(const struct shell *shell, size_t argc, char *argv[], bool acked)
+{
+	bool on_off = shell_model_str2bool(argv[1]);
+	uint32_t time = (argc >= 3) ? (uint32_t)strtol(argv[2], NULL, 0) : 0;
+	uint32_t delay = (argc == 4) ? (uint32_t)strtol(argv[3], NULL, 0) : 0;
+
+	if (!mod) {
+		if (!shell_model_first_get(BT_MESH_MODEL_ID_GEN_ONOFF_CLI, &mod)) {
+			return -ENODEV;
+		}
+	}
+
+	struct bt_mesh_onoff_cli *cli = mod->user_data;
+	struct bt_mesh_model_transition trans = { .time = time, .delay = delay };
+	struct bt_mesh_onoff_set set = { .on_off = on_off,
+					 .transition = (argc > 2) ? &trans : NULL };
+
+	if (acked) {
+		struct bt_mesh_onoff_status rsp;
+		int err = bt_mesh_onoff_cli_set(cli, NULL, &set, &rsp);
+
+		if (!err) {
+			shell_print(shell, "Present val: %d, target val: %d, rem time: %d",
+				    rsp.present_on_off, rsp.target_on_off, rsp.remaining_time);
+		}
+
+		return err;
+	} else {
+		return bt_mesh_onoff_cli_set_unack(cli, NULL, &set);
+	}
+}
+
+static int shell_onoff_set(const struct shell *shell, size_t argc, char *argv[])
+{
+	return onoff_set(shell, argc, argv, true);
+}
+
+static int shell_onoff_set_unack(const struct shell *shell, size_t argc, char *argv[])
+{
+	return onoff_set(shell, argc, argv, false);
+}
+
+static int shell_instance_get(const struct shell *shell, size_t argc, char *argv[])
+{
+	return shell_model_instances_get(shell, BT_MESH_MODEL_ID_GEN_ONOFF_CLI);
+}
+
+static int shell_instance_set(const struct shell *shell, size_t argc, char *argv[])
+{
+	uint8_t elem_idx = (uint8_t)strtol(argv[1], NULL, 0);
+
+	return shell_model_instance_set(shell, &mod, BT_MESH_MODEL_ID_GEN_ONOFF_CLI, elem_idx);
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(instance_cmds,
+	SHELL_CMD_ARG(set, NULL, "<elem_idx> ", shell_instance_set, 2, 0),
+	SHELL_CMD_ARG(get, NULL, NULL, shell_instance_get, 1, 0),
```

Comment:
After reading the instances code, there is a semantic difference between these two commands and they are not symmetrical. 

"instance set" would set the client instance to the one on the specified element. However,
"instance get" would not print the currently selected instance, but instead print the all the element indices where the given model is available. This interface could be improved as follows:

instance set : <leave as is>
instance get-cur: <tells what is the currently selected instance is> (or skip this command all together)
instance get-all: <does what the current 'instance get' does>

Reply:
Actually, is there a use case in `instance get-cur` command? If a user is not sure about the current model, call `instance set`.

Reply:
Also, using `mesh dst <local>` and `mesh get-comp` a user can get all model instances on the device. This removes the need in `instance get-all` commands for every model.

Reply:
So you think we could manage with only the `instance set` command? I have no strong opinion on the `instance get_cur` feature, but I think the `instance get-all` feature is a valid convenience tool. I'm not sure if the average user of the interface would have sufficient knowledge to know that it is possible to obtain the same information with the commands you mentioned.

Reply:
How to get the list of models on the node can be and should be described in a .rst file.

Reply:
I agree on the .rst suggestion. Regarding the `mesh get-comp` I see that this command currently only is included if the `CONFIG_BT_MESH_CFG_CLI` option is enabled. In its current state your I guess your suggestion would depend on the application containing a cfg_cli instance?

Reply:
Regarding instance get-cur: Yes, that is what is suggested in the parenthesis. You can skip it. Not much use anyway.

Reply:
> Regarding the `mesh get-comp` I see that this command currently only is included if the `CONFIG_BT_MESH_CFG_CLI` option is enabled. In its current state your I guess your suggestion would depend on the application containing a cfg_cli instance?

Ok, I see, missed that.

Reply:
May be "instance get-all" seems like a common feature and could be somehow inserted in the mesh shell itself, to enable query for a specific model. For example:
$ mesh instance get-all client mdl-onoff

Here, "mdl-onoff" would be an enumerated string for various models. This may be a bit of bloat.

I am leaning more towards Pavel's suggestion of using "mesh get-comp". This will save several bytes of code.  The help string at Line 93 could say:
> SHELL_CMD_ARG(set, NULL, "<elem_idx> \n use '$ mesh get-comp' to find elem_idx for this model", shell_instance_set, 2, 0),

This common help string ("<elem_idx> \n use '$ mesh get-comp' to find elem_idx for this model") could be placed in one common file and declared as extern const in this file to further minimize footprint for this and all other client shell modules

Reply:
@omkar3141 As I mentioned the `mesh get-comp` is dependent on the config client model as of today. If we are using a common mesh shell feature I think it should not have any such dependencies since that would decrease the value of the shell models greatly IMO. We could of course implement a simmilar feature in the `mesh` shell tree, but im not sure if that should be a separate task.

Reply:
The user needs a config client anyway. Otherwise one cannot change the publish settings. For a typical use case scenario, much of the functionality cannot be used if the user cannot configure remote servers' model settings or AppKey additions, etc. So, only client model shells will never be useful on their own. Users will almost always need a main mesh shell to be able to provision and configure remote devices.